pkg/morph/deploy/contracts.go

```diff
@@ -1,7 +1,410 @@
+					l.Error("failed to send transaction deploying the contract, will try again later", zap.Error(err))
+				}
+				continue
+			}
+
+			l.Info("transaction deploying the contract has been successfully sent, will wait for the outcome",
+				zap.Stringer("tx", txID), zap.Uint32("vub", vub),
+			)
+
+			deployTxMonitor.trackPendingTransactionsAsync(ctx, vub, txID)
+
+			continue
+		}
+
+		if onChainState.NEF.Checksum == prm.localNEF.Checksum {
+			// manifests may differ, but currently we should bump internal contract version
+			// (i.e. change NEF) to make such updates. Right now they are not supported due
+			// to dubious practical need
+			// Track https://github.com/nspcc-dev/neofs-contract/issues/340
+			l.Info("same local and on-chain checksums of the the contract NEF, update is not needed")
+
+			if !missingDomainName && !missingDomainRecord {
+				return onChainState.Hash, nil
+			}
+		} else if !alreadyUpdated {
+			// see comments same code block in updateNNSContract
+			l.Info("NEF checksums of the on-chain and local contract states differ, need an update")
+
+			versionOnChain, err := readContractOnChainVersion(prm.blockchain, onChainState.Hash)
+			if err != nil {
+				l.Error("failed to read on-chain version of the contract, will try again later", zap.Error(err))
+				continue
+			}
+
+			if v := localVersion.cmp(versionOnChain); v == -1 {
+				prm.logger.Info("local contract version is < than the on-chain one, update is not needed",
+					zap.Stringer("local", localVersion), zap.Stringer("on-chain", versionOnChain))
+				return onChainState.Hash, nil
+			} else if v == 0 {
+				return util.Uint160{}, fmt.Errorf("local and on-chain contracts have different NEF checksums but same version '%s'", versionOnChain)
+			}
+
+			extraUpdateArgs, err := prm.buildVersionedExtraUpdateArgs(versionOnChain)
+			if err != nil {
+				l.Error("failed to prepare build extra arguments for the contract update, will try again later",
+					zap.Stringer("on-chain version", versionOnChain), zap.Error(err))
+				continue
+			}
+
+			tx, err := committeeActor.MakeCall(onChainState.Hash, methodUpdate,
+				bLocalNEF, jLocalManifest, extraUpdateArgs)
+			if err != nil {
+				if isErrContractAlreadyUpdated(err) {
+					// note that we can come here only if local version is > than the on-chain one
+					// (compared above)
+					l.Info("the contract has already been updated")
+					if !missingDomainName && !missingDomainRecord {
+						return onChainState.Hash, nil
+					}
+					alreadyUpdated = true
+				} else {
+					l.Error("failed to make transaction updating the contract, will try again later", zap.Error(err))
+				}
+				continue
+			}
+
+			if updateTxMonitor.isPending() {
+				l.Info("previously sent Notary request updating the contract is still pending, will wait for the outcome")
+				continue
+			}
+
+			l.Info("sending new Notary request updating the contract...")
+
+			mainTxID, fallbackTxID, vub, err := committeeActor.Notarize(tx, nil)
```

Comment:
nothing: they all act on their own because update can't be synchronized in general. Do u think we should syns nonce/vub, and, if so, how?

Reply:
Update requires alphabet signature, therefore we use notary to collect signatures, but it has to have some stable parameters wrt this, for example, `nonce == epoch` and `vub == last_epoch_block + something`. Yeah, `epoch` can diverge too, but there is a big enough timeframe for it to work in general.

Reply:
>  it has to have some stable parameters wrt this

wanna clarify: is this necessary so as not to send multiple Notary requests with the same context from different nodes? when re-signing, then nonce and vub are known

Reply:
Yeah, we're likely to get some race here since local signers will always be different for different nodes. But this can be solved in an additional issue (reuse proxy for example, deploy it early in the sequence).

Reply:
ok i'm trying, will add commit above, will squash later if needed

Reply:
done, check
* proxy 0b21b3a858c51a96fa7e2e5f0fa990e1f47fde4e
* epochs 3426dbc26310b7c1ca256b4d80383f2faf3daa95